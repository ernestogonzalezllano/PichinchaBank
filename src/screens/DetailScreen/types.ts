import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../types/screensTypes';

export type DetailScreenProps = NativeStackScreenProps<
  RootStackParamList,
  'Detail'
>;
