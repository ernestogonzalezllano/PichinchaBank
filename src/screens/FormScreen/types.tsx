import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../types/screensTypes';
import {GapsNames} from '../../types/api';

export type FormScreenProps = NativeStackScreenProps<
  RootStackParamList,
  'Form'
>;

export type FormGaps = {
  [key in GapsNames]: string | undefined;
};
