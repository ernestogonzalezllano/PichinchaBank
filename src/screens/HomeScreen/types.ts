import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../types/screensTypes';

export type HomeProps = NativeStackScreenProps<RootStackParamList, 'Home'>;
