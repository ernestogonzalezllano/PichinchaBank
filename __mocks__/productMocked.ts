export const productMocked = {
  id: '12345',
  date_release: '2024-11-11',
  date_revision: '2025-11-11',
  description: 'test react testing library',
  logo: 'https://www.visa.com.ec/dam/VCOM/regional/lac/SPA/Default/Pay%20With%20Visa/Tarjetas/visa-signature-400x225.jpg',
  name: 'test test',
};
