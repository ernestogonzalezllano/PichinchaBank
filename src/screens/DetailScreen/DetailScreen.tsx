import React from 'react';
import {StyleSheet, Text, View} from 'react-native';
import {DetailRow} from '../../components/DetailRow';
import {LogoRow} from '../../components/LogoRow';
import {ButtonTouchable} from '../../components/ButtonTouchable';
import {GAPS} from '../../constants';
import {isValidDateString} from '../../utils/validators';
import {DetailScreenProps} from './types';
import {Modal} from './components/Modal';
import {useDetailScreen} from './hooks/useDetailScreen';

export const DetailScreen = ({
  route: {params: product},
  navigation: {navigate, popToTop},
}: DetailScreenProps) => {
  const {openModal, setOpenModal, refetch, loading, error, data} =
    useDetailScreen({product});
  return (
    <View style={styles.container}>
      <View style={styles.detailContainer}>
        <View>
          <Text style={styles.textId}>ID:{product.id}</Text>
          <Text>Informaci√≥n extra</Text>
        </View>
        <View style={styles.infoContainer}>
          {Object.keys(GAPS).map(gap => {
            const isDate = isValidDateString(
              product[gap as keyof typeof product],
            );
            if (isDate) {
              product = {...product, [gap]: isDate};
            }
            const props = {
              key: gap,
              value: product[gap as keyof typeof product],
              gapName: GAPS[gap as keyof typeof GAPS].label,
            };
            return gap === 'logo' ? (
              <LogoRow {...props} />
            ) : (
              <DetailRow {...props} />
            );
          })}
        </View>
      </View>

      <View style={styles.footerContainer}>
        <ButtonTouchable
          label="Editar"
          type="secondary"
          onPress={() => {
            navigate('Form', product);
          }}
        />
        <ButtonTouchable
          label="Eliminar"
          type="danger"
          onPress={() => setOpenModal(true)}
        />
      </View>
      {openModal && (
        <Modal
          {...{
            loading,
            refetch,
            setOpenModal,
            error,
            data,
            handleGoToHome: () => popToTop(),
          }}
        />
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {flex: 1},
  detailContainer: {flex: 1},
  headerContainer: {},
  infoContainer: {marginTop: 10},
  textId: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  footerContainer: {
    alignSelf: 'flex-end',
    width: '100%',
    gap: 10,
  },
});
