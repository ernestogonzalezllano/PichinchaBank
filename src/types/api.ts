export type ApiRoutes = '/bp/products' | '/bp/products/verification';

export type GapsNames =
  | 'id'
  | 'name'
  | 'description'
  | 'logo'
  | 'date_release'
  | 'date_revision';

export interface Product {
  name: string;
  description: string;
  logo: string;
  date_release: string;
  date_revision: string;
  id: string;
}
