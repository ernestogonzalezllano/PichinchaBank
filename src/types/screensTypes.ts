import {Product} from './api';

export type RootStackParamList = {
  Home: undefined;
  Detail: Product;
  Form?: Product;
};
