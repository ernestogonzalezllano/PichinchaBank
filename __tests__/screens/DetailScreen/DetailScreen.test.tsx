import * as React from 'react';
import {
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react-native';
import {NavigationContainer} from '@react-navigation/native';

import {DetailScreen} from '../../../src/screens/DetailScreen/DetailScreen';
import {DetailScreenProps} from '../../../src/screens/DetailScreen/types';

import {productMocked as originalProductMocked} from '../../../__mocks__/productMocked';

const productMocked = {
  ...originalProductMocked,
};
const navigate = jest.fn();

const props = {
  navigation: {
    navigate,
  },
  route: {
    params: productMocked,
  },
} as unknown as DetailScreenProps;

describe('DetailScreen', () => {
  test('Should match with snapshot', async () => {
    const component = renderComponent();
    expect(component).toMatchSnapshot();
  });
  test('Should go to ScreenForm when "Editar" button is pressed', async () => {
    renderComponent();
    fireEvent.press(screen.getByText('Editar'));
    expect(navigate).toHaveBeenCalledWith('Form', {
      ...productMocked,
      date_release: '2024-11-10',
      date_revision: '2025-11-10',
    });
  });
  test('Should close Modal when "Cancelar" button is pressed', async () => {
    await renderModal();
    fireEvent.press(screen.getByText('Cancelar'));
    await waitFor(() => {
      expect(screen.queryByText('Estas seguro de eliminar?')).toBeNull();
    });
  });
  test('Should close Modal when CloseIcon is pressed', async () => {
    await renderModal();
    fireEvent.press(screen.getByTestId('CloseIcon'));
    await waitFor(() => {
      expect(screen.queryByText('Estas seguro de eliminar?')).toBeNull();
    });
  });
});

const renderComponent = () => {
  const component = render(
    <NavigationContainer>
      <DetailScreen {...props} />
    </NavigationContainer>,
  );
  return component;
};

const renderModal = async () => {
  renderComponent();
  fireEvent.press(screen.getByText('Eliminar'));
  await waitFor(() => {
    expect(screen.getByText('Estas seguro de eliminar?')).toBeOnTheScreen();
  });
};
