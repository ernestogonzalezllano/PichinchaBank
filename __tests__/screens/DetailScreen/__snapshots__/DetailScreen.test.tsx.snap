// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailScreen Should match with snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <Text
        style={
          {
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        ID:
        12345
      </Text>
      <Text>
        Informaci贸n extra
      </Text>
    </View>
    <View
      style={
        {
          "marginTop": 10,
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 10,
          }
        }
      >
        <Text
          style={{}}
        >
          ID
        </Text>
        <Text
          style={
            {
              "flex": 1,
              "fontWeight": "bold",
              "textAlign": "right",
            }
          }
        >
          12345
        </Text>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 10,
          }
        }
      >
        <Text
          style={{}}
        >
          Nombre
        </Text>
        <Text
          style={
            {
              "flex": 1,
              "fontWeight": "bold",
              "textAlign": "right",
            }
          }
        >
          test test
        </Text>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 10,
          }
        }
      >
        <Text
          style={{}}
        >
          Descripci贸n
        </Text>
        <Text
          style={
            {
              "flex": 1,
              "fontWeight": "bold",
              "textAlign": "right",
            }
          }
        >
          test react testing library
        </Text>
      </View>
      <View
        style={
          {
            "marginVertical": 10,
          }
        }
      >
        <Text>
          Logo
        </Text>
        <Image
          source={
            {
              "uri": "https://www.visa.com.ec/dam/VCOM/regional/lac/SPA/Default/Pay%20With%20Visa/Tarjetas/visa-signature-400x225.jpg",
            }
          }
          style={
            {
              "height": 100,
              "marginTop": 5,
              "resizeMode": "center",
              "width": "100%",
            }
          }
        />
      </View>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 10,
          }
        }
      >
        <Text
          style={{}}
        >
          Fecha liberaci贸n
        </Text>
        <Text
          style={
            {
              "flex": 1,
              "fontWeight": "bold",
              "textAlign": "right",
            }
          }
        >
          2024-11-10
        </Text>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 10,
          }
        }
      >
        <Text
          style={{}}
        >
          Fecha revisi贸n
        </Text>
        <Text
          style={
            {
              "flex": 1,
              "fontWeight": "bold",
              "textAlign": "right",
            }
          }
        >
          2025-11-10
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "alignSelf": "flex-end",
        "gap": 10,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#eaecf3",
          "borderRadius": 5,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          {
            "backgroundColor": "#eaecf3",
            "fontWeight": "bold",
          }
        }
      >
        Editar
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#d50708",
          "borderRadius": 5,
          "color": "#fff",
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          {
            "backgroundColor": "#d50708",
            "color": "#fff",
            "fontWeight": "bold",
          }
        }
      >
        Eliminar
      </Text>
    </View>
  </View>
</View>
`;
